File: plugin/src/main/java/io/github/kdy05/physicalFighters/ability/impl/Kimimaro.java
package io.github.kdy05.physicalFighters.ability.impl;

import io.github.kdy05.physicalFighters.ability.Ability;
import io.github.kdy05.physicalFighters.ability.AbilitySpec;
import io.github.kdy05.physicalFighters.game.EventManager;
import io.github.kdy05.physicalFighters.util.BaseItem;
import io.github.kdy05.physicalFighters.util.EventData;
import io.github.kdy05.physicalFighters.util.PotionEffectFactory;
import org.bukkit.Material;
import org.bukkit.entity.LivingEntity;
import org.bukkit.event.Event;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;

import java.util.UUID;

public class Kimimaro extends Ability implements BaseItem {
    public Kimimaro(UUID playerUuid) {
        super(AbilitySpec.builder("키미마로", Type.PassiveManual, Rank.SS)
                .guide("뼈다귀로 상대를 공격할 시에 강한 대미지를 주고,",
                        "40% 확률로 상대에게 5초간 독 효과를 겁니다.")
                .build(), playerUuid);
    }

    @Override
    public void registerEvents() {
        EventManager.registerEntityDamageByEntity(new EventData(this, 0));
    }

    @Override
    public int checkCondition(Event event, int CustomData) {
        if (CustomData == 0) {
            EntityDamageByEntityEvent event0 = (EntityDamageByEntityEvent) event;
            if (isOwner(event0.getDamager()) && isValidItem(Material.BONE)
                    && event0.getEntity() instanceof LivingEntity)
                return 0;
        }
        return -1;
    }

    @Override
    public void applyEffect(Event event, int CustomData) {
        if (CustomData == 0) {
            EntityDamageByEntityEvent event0 = (EntityDamageByEntityEvent) event;
            event0.setDamage(event0.getDamage() * 7);
            LivingEntity entity = (LivingEntity) event0.getEntity();
            if (Math.random() < 0.4D)
                entity.addPotionEffect(PotionEffectFactory.createPoison(20 * 5, 0));
        }
    }

    @NotNull
    @Override
    public ItemStack[] getBaseItem() {
        return new ItemStack[] {
                new ItemStack(Material.BONE, 1)
        };
    }

    @NotNull
    @Override
    public String getItemName() {
        return "뼈다귀";
    }

}

File: abilities/src/main/kotlin/io/github/kdy05/physicalFighters/ability/impl/Kimimaro.kt
package io.github.kdy05.physicalFighters.ability.impl

import io.github.kdy05.abilityAPI.ability.AbilityMeta
import io.github.kdy05.abilityAPI.rank.Rank
import io.github.kdy05.abilityAPI.skill.SkillContext
import io.github.kdy05.physicalFighters.ability.PFAbility
import io.github.kdy05.physicalFighters.util.PotionEffectFactory
import org.bukkit.Material
import org.bukkit.entity.LivingEntity
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack

@AbilityMeta(
    name = "키미마로",
    rank = Rank.SS,
    guide = [
        "뼈다귀로 상대를 공격할 시에 강한 대미지를 주고,",
        "40% 확률로 상대에게 5초간 독 효과를 겁니다."
    ]
)
class Kimimaro(owner: Player, context: SkillContext) : PFAbility(owner, context) {
    override fun buildSkills() = listOf(object : BaseItemSkill() {
        override val baseItems = arrayOf(ItemStack(Material.BONE, 1))
        override val itemDisplayName = "뼈다귀"

        override fun register() {
            registerItemEvents()
            onEntityDamage { e ->
                if (owner.inventory.itemInMainHand.type != Material.BONE) return@onEntityDamage
                val entity = e.entity as? LivingEntity ?: return@onEntityDamage
                e.damage = owner.attackCooldown.toDouble() * 7.0
                if (Math.random() < 0.4) {
                    entity.addPotionEffect(PotionEffectFactory.createPoison(100, 0))
                }
            }
        }
    })
}
